libraries/AP_RangeFinder/AP_RangeFinder_VL53L1X.cpp

```diff
@@ -0,0 +1,474 @@
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+  driver for ST VL53L1X lidar
+
+  Many thanks to Pololu, https://github.com/pololu/vl53l1x-arduino and
+  the ST example code
+ */
+#include "AP_RangeFinder_VL53L1X.h"
+
+#include <utility>
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/utility/sparse-endian.h>
+#include <stdio.h>
+
+extern const AP_HAL::HAL& hal;
+
+/*
+   The constructor also initializes the rangefinder. Note that this
+   constructor is not called until detect() returns true, so we
+   already know that we should setup the rangefinder
+*/
+AP_RangeFinder_VL53L1X::AP_RangeFinder_VL53L1X(RangeFinder::RangeFinder_State &_state, AP_HAL::OwnPtr<AP_HAL::I2CDevice> _dev)
+    : AP_RangeFinder_Backend(_state)
+    , dev(std::move(_dev)) {}
+
+/*
+   detect if a VL53L1X rangefinder is connected. We'll detect by
+   trying to take a reading on I2C. If we get a result the sensor is
+   there.
+*/
+AP_RangeFinder_Backend *AP_RangeFinder_VL53L1X::detect(RangeFinder::RangeFinder_State &_state, AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev)
+{
+    if(!dev){
+      	return nullptr;
+    }
+        AP_RangeFinder_VL53L1X *sensor
+            = new AP_RangeFinder_VL53L1X(_state, std::move(dev));
+
+        if (!sensor) {
+            delete sensor;
+            return nullptr;
+        }
+
+        sensor->dev->get_semaphore()->take_blocking();
+        
+        if (!sensor->check_id() || !sensor->init()) {
+            sensor->dev->get_semaphore()->give();
+            delete sensor;
+            return nullptr;
+        }
+
+        sensor->dev->get_semaphore()->give();
+        
+        return sensor;
+}
+
+// check sensor ID registers
+bool AP_RangeFinder_VL53L1X::check_id(void)
+{
+    uint8_t v1, v2;
+    v1 = read_register(0x010F);
+    v2 = read_register(0x0110);
+
+    if ((v1 != 0xEA) ||
+        (v2 != 0xCC)) {
+        return false;
+    }
+    printf("Detected VL53L1X on bus 0x%x\n", dev->get_bus_id());
+    return true;
+}
+
+bool AP_RangeFinder_VL53L1X::init()
+{
+    // setup for 2.8V operation
+    write_register(PAD_I2C_HV__EXTSUP_CONFIG,
+                   read_register(PAD_I2C_HV__EXTSUP_CONFIG) | 0x01);
+
+    // store oscillator info for later use
+    fast_osc_frequency = read_register16(OSC_MEASURED__FAST_OSC__FREQUENCY);
+    osc_calibrate_val = read_register16(RESULT__OSC_CALIBRATE_VAL);
+    
+    // static config
+    write_register16(DSS_CONFIG__TARGET_TOTAL_RATE_MCPS, TargetRate); // should already be this value after reset
```

Comment:
I am thinking of using a splitter. I think that it is better to judge processing results by considering bus noise and signal slowing.
and other statements.

Reply:
Hi @muramura,
Sorry but I don't understand what you mean by using splitter? Can you give me an example of what you think it should be?